### PR TITLE
🐛 [#4199] Delete authinfo from session after it's stored on submission 

### LINKED ISSUE #4199: Session persistance causes login choice te be ignored
### Product versie / Product version

2.6.x

### Omschrijf het probleem / Describe the bug

When form A is a form with authentication and i start form B without authentication, form B will ignore this and also be authenticated. 
Complete with prefill and auth_* object being present in the submission. 

This behavior does not seem correct, my choice when starting a form should not be ignored and my authentication object should not be present in form B.

### Stappen om te reproduceren / Steps to reproduce

Create 2 forms that share authentication options
Open both front-ends of the forms
Login to first form using open "Inloggen met Demo BAN (test)"
Login to seconde form not using "Inloggen met Demo BAN (test)" but "Start form" 
Check both submissions, the should contain the same Authentication details:
image

### Verwacht gedrag / Expected behavior

_No response_

### Screen resolution

None

### Device

None

### OS

None

### Browser

_No response_

### PR DESCRIPTION
**Pull request by stevenbal**
Closes #4199 partly
Related PR: https://github.com/open-formulieren/open-forms-sdk/pull/688

**Changes**

* Delete authinfo from session after it's stored on submission 
* Add explicit `anonymous` option to submission endpoint

TODO:
- [x] add test for anonymous attribute
- [x] check where `FORM_AUTH_SESSION_KEY` is used
  - [x] set_auth_attribute_on_session
  - [x] set_cosign_data_on_submission (cosign also uses is_authenticated_with_an_allowed_plugin)
  - [x] ResumeFormMixin: is_auth_data_correct / get_redirect_url -> user is always forced to login again (is this an issue?)
      - [x] is_authenticated_with_plugin / meets_plugin_requirements -> run after relogging, so this is fine
      - [x] EHerkenningAuthentication.check_requirements -> used in meets_plugin_requirements which is only used for resume flow
      - [x] VerifyChangeAppointmentLinkView.custom_submission_modifications -> only used in resume flow
- [x] expose function to remove auth info from session in service.py
- [x] call this function at end of add_submission_to_session and other functions that rely on this attr (?)
- [x] update branch in CI of token-exchange lib to see if it passes: https://github.com/open-formulieren/open-forms-ext-token-exchange/pull/16

**Checklist**

Check off the items that are completed or not relevant.

- Impact on features

  - [x] Checked copying a form
  - [x] Checked import/export of a form
  - [x] Config checks in the configuration overview admin page
  - [x] Problem detection in the admin email digest is handled

- Release management

  - [x] I have labelled the PR as "needs-backport" accordingly

- I have updated the translations assets (you do NOT need to provide translations)

  - [x] Ran `./bin/makemessages.sh`
  - [x] Ran `./bin/compilemessages_js.sh`

- Commit hygiene

  - [x] Commit messages refer to the relevant Github issue
  - [x] Commit messages explain the "why" of change, not the how

### CODE REVIEW

File: src/openforms/authentication/signals.py
Comment (stevenbal) [reply src/openforms/authentication/signals.py:128]: I looked into this, but I'm not entirely sure how I can test it locally (because it seems this is only implemented for BRP?). Judging from the explanation in the [docs](https://open-forms.readthedocs.io/en/latest/developers/extensions.html#keycloak-token-exchange-extension), I don't think it will be an issue because the `request.session[FORM_AUTH_SESSION_KEY]` is deleted after the `submission_start` signal (step 2) is sent and isn't mentioned outside of step 1.

I do however wonder if this won't cause issues for several utils in `openforms.authentication.utils` (e.g. `is_authenticated_with_plugin`), because these utils do rely on `request.session[FORM_AUTH_SESSION_KEY]`.

**Comment by stevenbal** [comment]
@sergei-maertens I'm not 100% sure if this needs backport? What is our policy for when to backport bugfixes?

**Comment by stevenbal** [comment]
@sergei-maertens I checked the usage of the session auth info and I didn't notice any problems if we remove it from the session, the only thing that changes is that when resuming a form, the user will always be forced to login again even if it's within the same session (because the authinfo isn't persisted on the session anymore)

**Comment by sergei-maertens** [comment]
> @sergei-maertens I'm not 100% sure if this needs backport, what is our policy for when to backport bugfixes?

We typically label the ticket as such during refinement, based on how big the impact is and whether a workaround exists or not.

I think that the odds that a end user is filling out multiple forms at the same time where one is authenticated and the other one not are pretty low, so I don't think it warrants a backport. There also doesn't seem to be a stakeholder (except for us) in this, so it doesn't seem to be a big problem.

**Comment by sergei-maertens** [comment]
> @sergei-maertens I checked the usage of the session auth info and I didn't notice any problems if we remove it from the session, the only thing that changes is that when resuming a form, the user will always be forced to login again even if it's within the same session (because the authinfo isn't persisted on the session anymore)

I think that's a good side-effect of this patch, it fits better with the expectations that resuming always forces you to re-authenticate and creates less different authentication flows for the same "mechanism".

File: src/openforms/authentication/tests/test_signals.py
Comment (sergei-maertens) [inline src/openforms/authentication/tests/test_signals.py:145]: is the absence of `anonymous=True` deliberate here?

File: src/openforms/authentication/tests/test_utils.py
Comment (sergei-maertens) [inline src/openforms/authentication/tests/test_utils.py:88]:
```suggestion
```
`APIRequestFactory` already produces a `Request` instance, no?

I'd expect your type checker to complain about this too btw

File: src/openforms/submissions/api/viewsets.py
Comment (sergei-maertens) [inline src/openforms/submissions/api/viewsets.py:172]: maybe provide a default/initial on the serializer field so you can do `serializer.validated_data["anonymous"]`? That's a bit more straight-forward.

**Review by sergei-maertens — CHANGES_REQUESTED** [review]
I'd like to see a test with explicit anonymous start in `openforms.submissions.tests.test_start_submission`.

I don't put too much value on the unit tests of the signal, I'm mostly interested in the integration tests for the `POST` call that those behave as expected, so please add a test that sets up an authentication state and creates an anonymous submission and make assertions on that created submission.

File: src/openforms/authentication/tests/test_utils.py
Comment (stevenbal) [reply src/openforms/authentication/tests/test_utils.py:88]: I thought so too, that is why I used `APIRequestFactory` instead of `RequestFactory` but strangely enough my type checker complains about this if I don't explicitly cast it:

Screenshot from 2024-05-23 14-23-20

File: src/openforms/authentication/tests/test_signals.py
Comment (stevenbal) [reply src/openforms/authentication/tests/test_signals.py:145]: I made this test before I added the `anonymous` flag I think, I'll add it when I move these tests to the `start_submission` tests

File: src/openforms/submissions/api/serializers.py
Comment (stevenbal) [inline src/openforms/submissions/api/serializers.py:156]: I think this is an issue with django and typing, but at this point I don't know if I can trust my type checker :stuck_out_tongue: 
Screenshot from 2024-05-23 15-23-06

**Comment by stevenbal** [comment]
Is this a known flaky test? @sergei-maertens https://github.com/open-formulieren/open-forms/actions/runs/9208977538/job/25332413671?pr=4271#step:7:3347

File: src/openforms/submissions/api/serializers.py
Comment (sergei-maertens) [reply src/openforms/submissions/api/serializers.py:156]: yeah it is, maybe that the drf-stubs help with this, but otherwise you can indeed just `type: ignore` it :)

**Comment by sergei-maertens** [comment]
> Is this a known flaky test? @sergei-maertens https://github.com/open-formulieren/open-forms/actions/runs/9208977538/job/25332413671?pr=4271#step:7:3347

Yeah, sporadically it crashes. hypothesis tends to reveal those cases...

**Review by sergei-maertens — APPROVED** [review]
I'll leave it up to you to hit the merge button if this is ready to be merged in your opinion :)

File: src/openforms/authentication/tests/test_utils.py
Comment (stevenbal) [reply src/openforms/authentication/tests/test_utils.py]: I'll add `AnyRequest` :+1:

Comment (stevenbal) [reply src/openforms/authentication/tests/test_utils.py]: So this means my type checker isn't messed up? :smile:

**Comment by stevenbal** [comment]
> I'll leave it up to you to hit the merge button if this is ready to be merged in your opinion :)

I pushed the typing changed and will merge it once CI passes

**Pull request merged**
